crates/ruff_python_parser/src/parser/statement.rs

```diff
@@ -987,191 +987,372 @@ impl<'src> Parser<'src> {
+                // with (item := 10): ...
+                //
+                // # Multiple items
+                // with (item1, item2 := 10): ...
+                // ```
+                TokenKind::ColonEqual => {
+                    let named_expr = self.parse_named_expression(parsed_expr.expr, start);
+
+                    // For example: `with (item := 10 as foo): ...`
+                    if self.at(TokenKind::As) {
+                        self.add_error(
+                            ParseErrorType::OtherError(
+                                "unparenthesized named expression cannot be used here".to_string(),
+                            ),
+                            named_expr.range(),
+                        );
+                    }
 
-        let optional_vars = if self.eat(TokenKind::As) {
-            let mut target = self.parse_conditional_expression_or_higher();
+                    Expr::Named(named_expr).into()
+                }
+                TokenKind::Async | TokenKind::For => {
+                    let generator_expr =
+                        if let WithItemParsingState::AmbiguousLparFirstItem(lpar_start) = state {
+                            // The parser is at the first with item after the ambiguous `(` token.
+                            // For example:
+                            //
+                            // ```python
+                            // with (x for x in range(10)): ...
+                            // with (x for x in range(10)), item: ...
+                            // ```
+                            used_ambiguous_lpar = true;
+                            self.parse_generator_expression(parsed_expr.expr, lpar_start, true)
+                        } else {
+                            // For better error recovery. We would not take this path if the
+                            // expression was parenthesized as it would be parsed as a generator
+                            // expression by `parse_conditional_expression_or_higher`.
+                            //
+                            // ```python
+                            // # This path will be taken for
+                            // with (item, x for x in range(10)): ...
+                            //
+                            // # This path will not be taken for
+                            // with (item, (x for x in range(10))): ...
+                            // ```
+                            self.parse_generator_expression(parsed_expr.expr, start, false)
+                        };
+
+                    if !generator_expr.parenthesized {
+                        self.add_error(
+                            ParseErrorType::OtherError(
+                                "unparenthesized generator expression cannot be used here"
+                                    .to_string(),
+                            ),
+                            generator_expr.range(),
+                        );
+                    }
```

Comment:
Nit: I think we can move the `!generator_expr.parenthesized` into the else branch
```suggestion
                        } else {
                            // For better error recovery. We would not take this path if the
                            // expression was parenthesized as it would be parsed as a generator
                            // expression by `parse_conditional_expression_or_higher`.
                            //
                            // ```python
                            // # This path will be taken for
                            // with (item, x for x in range(10)): ...
                            //
                            // # This path will not be taken for
                            // with (item, (x for x in range(10))): ...
                            // ```
                            let generator_expr = self.parse_generator_expression(parsed_expr.expr, start, false);
                            self.add_error(
                                ParseErrorType::OtherError(
                                    "unparenthesized generator expression cannot be used here"
                                        .to_string(),
                                ),
                                generator_expr.range(),
                            );
                            generator_expr
                        };
```